### PR TITLE
Allow the serialization of non-collection types implementing IEnumerable

### PR DESCRIPTION
**Pull request by samcragg**
This is designed to fix issue #169 

I've added a compatibility option to disable the new behaviour (`AlwaysAssumeCollections`) and all the unit tests to pass with the refactored code, plus there's an additional one showing the desired behaviour.

Since this is my first pull request let me know if there's anything you need me to do

### CODE REVIEW

File: test/MsgPack.UnitTest/Serialization/NonCollectionIEnumerableTest.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+				var obj = new NonCollectionType { Property = 123 };
+				serializer.Pack(buffer, obj);
+
+				buffer.Position = 0;
+
+				var clone = serializer.Unpack( buffer );
+				Assert.That( clone.Property, Is.EqualTo( 123 ) );
+			}
+		}
+
```
Comment (yfakariya) [inline test/MsgPack.UnitTest/Serialization/NonCollectionIEnumerableTest.cs]: Can I move this test to AutoMessagePackSerializerTest.ttinclude and regenerate {Array|Map}{Field|Generation|Reflection}Based.tt? It is not documented how to do it, so I will do it after merge.

File: test/MsgPack.UnitTest/Serialization/NonCollectionIEnumerableTest.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		[Serializable]
```
Comment (yfakariya) [inline test/MsgPack.UnitTest/Serialization/NonCollectionIEnumerableTest.cs]: `[Serializable]` is redundant.

**Comment by samcragg** [comment]
I've just moved the test to AutoMessagePackSerializerTest.ttinclude (and removed the `[Serializable]` as requested) and it looks like some of the new generated tests fail (I think I've focused on one area that was causing me pain but the unit tests cover far more edge cases that need fixing first).

Let me investigate and I'll add a new commit with the changes before anything gets merged

**Comment by yfakariya** [comment]
Thank you for great work, and sorry I didn't tell you detailed instructions and wasted your time. We have some rules to work the AutoMessagePackSerializer tests, so you must fail without psychics. I failed to consider your passion.
(I don't write the instructions down here, because I don't want to uae your time anymore)
I will merge your PR and then do remained works for testing. And, of course, I will add developer guide for cobtribution.

File: src/MsgPack/Serialization/ReflectionExtensions.cs [hunk 3]
```diff
@@ -199,6 +199,13 @@ public static CollectionTraits GetCollectionTraits( this Type source, Collection
 					);
 			}
 
+			// If the type is an interface then a concrete collection has to be
+			// made for it if the interface is a collection type
+			if ( source.IsInterface || forceCollection )
```
Comment (yfakariya) [inline src/MsgPack/Serialization/ReflectionExtensions.cs]: Use `GetIsInterface()` internal extension method instead of `IsInterface` because `Type` of WinRT/UWP/.NET Core does not declare `IsInterface` property.

File: src/MsgPack/Serialization/SerializationCompatibilityOptions.cs [hunk 4]
```diff
@@ -126,6 +126,43 @@ public bool IgnorePackabilityForCollection
...
+		/// </summary>
+		/// <value>
+		///		<c>true</c> if serializer generator should serialize a type implementing IEnumerable as only a collection; otherwise, <c>false</c>. The default is <c>false</c>.
+		/// </value>
+		/// <remarks>
+		///		Historically, MessagePack for CLI always tried to serialize any type that implemented IEnumerable as a collection, throwing an exception
+		///		if an Add method could not be found. However, for types that implement IEnumerable but don't have an Add method the generator will now
+		///		serialize the type as a non-collection type. To restore the old behavior for backwards compatibility, set this option to <c>false</c>.
+		/// </remarks>
+		public bool AlwaysAssumeCollections
```
Comment (yfakariya) [inline src/MsgPack/Serialization/SerializationCompatibilityOptions.cs]: I think the meaning of this property is opposite.
If the serializer generator "always assume collections", thus it also assume "collection should implement Add".
This name should be `AllowEnumerableObject`(alternative name is welcome).

**Comment by yfakariya** [comment]
I'm trying to add tests for error cases, so I found some extra issues on your PR.
Sorry for taking back what I said, please check my comments.
In addition, I will fix and push some bugs they prevent your fix working correctly in an hour.

Here are the instructions and requests to testing:

1. Please add test following cases:
    * Non-generic version of `NonCollectionType`
    * If `AlwaysAssumeCollections` is not set to default, `SerializationException` is thrown. This case should also be added for non generic variant.
1. Put the `NonCollectionType` and non generic counter parts to `AutoMessagePackSerializerTest.Types.tt`.
1. Go to `PreGeneratedSerializerActivator.Types.tt`, put `result.Add( typeof( ... ) );` in the end of the method (after `#endif`) for both of generic one and non generic one.
1. Go to `PreGenerateSerializerGenerator.cs`, remove comment out of the `[Test]`. And execute the `GenerateFiles` method, then `MessagePack_Serialization_xxx.cs` (xxx is type name of generic one and non generic one) will be generated in `%SystemDrive%\temp-gen`. Move the files to `MsgPack.UnitTest\gen` folder.
1. Open MsgPack.compat.sln and go to MsgPack.UnitTest.Net35 project.
1. Run `PreGenerateSerializerGenerator.GenerateFiles()` test. New files will  be generated in `%SystemDrive%\temp-gen35`. Move the files to `MsgPackUnitTest.Net35\gen` folder.

File: src/MsgPack/Serialization/SerializerGenerator.cs [hunk 5]
```diff
@@ -557,7 +557,7 @@ protected override ISerializerCodeGenerationContext CreateGenerationContext( Ser
 
 			protected override Func<Type, ISerializerCodeGenerator> CreateGeneratorFactory()
 			{
-				return type => new AssemblyBuilderSerializerBuilder( type, type.GetCollectionTraits( CollectionTraitOptions.Full ) );
+				return type => new AssemblyBuilderSerializerBuilder( type, type.GetCollectionTraits( CollectionTraitOptions.Full, forceCollection: true ) );
```
Comment (yfakariya) [inline src/MsgPack/Serialization/SerializerGenerator.cs]: `forceCollection` should be configurable here.

File: src/MsgPack/Serialization/SerializerGenerator.cs [hunk 6]
```diff
@@ -577,7 +577,7 @@ protected override ISerializerCodeGenerationContext CreateGenerationContext( Ser
 
 			protected override Func<Type, ISerializerCodeGenerator> CreateGeneratorFactory()
 			{
-				return type => new CodeDomSerializerBuilder( type, type.GetCollectionTraits( CollectionTraitOptions.Full ) );
+				return type => new CodeDomSerializerBuilder( type, type.GetCollectionTraits( CollectionTraitOptions.Full, forceCollection: true ) );
```
Comment (yfakariya) [inline src/MsgPack/Serialization/SerializerGenerator.cs]: `forceCollection` should be configurable here too.

File: src/MsgPack/Serialization/SerializerGenerator.cs [hunk 7]
```diff
@@ -476,7 +476,7 @@ public IEnumerable<SerializerCodeGenerationResult> Generate( IEnumerable<Type> t
 
 			private static IEnumerable<Type> ExtractElementTypes( SerializationContext context, ISerializerGeneratorConfiguration configuration, Type type )
 			{
-				if ( !SerializationTarget.BuiltInSerializerExists( configuration, type, type.GetCollectionTraits( CollectionTraitOptions.None ) ) )
+				if ( !SerializationTarget.BuiltInSerializerExists( configuration, type, type.GetCollectionTraits( CollectionTraitOptions.None, forceCollection: true ) ) )
```
Comment (yfakariya) [inline src/MsgPack/Serialization/SerializerGenerator.cs]: `forceCollection` should be configurable here.

**Comment by yfakariya** [comment]
Thank you for great work! I merged with a few tweaking.
